### PR TITLE
[IMP] Use importlib find_spec() instead of find_module()

### PR DESCRIPTION
**Pull request by rousseldenis**
find_module() has been deprecated since 3.4:

https://github.com/python/cpython/blob/05c28b08f6e2fc8782472b026c98a3fdd61a2ba9/Lib/importlib/_bootstrap.py#L1347

Warning was added from 3.10:

https://github.com/python/cpython/blob/f91dfdf5ff9f68a4b012e1b70ab9997c6dc1542d/Lib/importlib/_bootstrap.py#L764

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by rousseldenis** [comment]
@rco-odoo

**Comment by Julien00859** [comment]
I'm +1 we change `find_module` for `find_spec` but we are trying to remove those hooks in #106054 so let's wait that other PR a bit before taking a decision here

**Comment by rousseldenis** [comment]
@Julien00859 It seems #106054 was closed before merge. @xmo-odoo Maybe this one is a good alternative

**Comment by marielejeune** [comment]
LGTM

File: odoo/modules/module.py [hunk 1]
```diff
@@ -8,15 +8,16 @@
...
 import warnings
-from os.path import join as opj, normpath
+from os.path import join as opj
+from os.path import normpath
 
 import odoo
-import odoo.tools as tools
 import odoo.release as release
+import odoo.tools as tools
+import pkg_resources
```
Comment (Julien00859) [inline odoo/modules/module.py:20]: I know the imports are ill-sorted but our policy is to sort them only when we actually add/remove new imports for better compatibility with our @fw-bot, can you please restore the original code?

**Review by Julien00859 — CHANGES_REQUESTED** [review]
Shouldn't we remove the `find_module` bits too then?

Can you change the title to `[FIX] core: importlib find_module is deprecated` and elaborate on the commit message, explaining the rationals and giving relevant links to the documentation? I'm speaking about the commit message, not the PR message. Only the commit message is accessible within the git history. Also the link <https://github.com/python/cpython/blob/main/Lib/importlib/_bootstrap.py#L739> doesn't seem to reference the correct line of code anymore, as a rule of thumb it is better to copy the *permanent* link (it creates a link to the current commit instead of to the current branch)

image

File: odoo/modules/module.py (hunk 1 above)
Comment (rousseldenis) [reply odoo/modules/module.py:20]: Yes indeed, my IDE changed them automatically.

**Comment by rousseldenis** [comment]
> Shouldn't we remove the `find_module` bits too then?
> 
> Can you change the title to `[FIX] core: importlib find_module is deprecated` and elaborate on the commit message, explaining the rationals and giving relevant links to the documentation? I'm speaking about the commit message, not the PR message. Only the commit message is accessible within the git history. Also the link https://github.com/python/cpython/blob/main/Lib/importlib/_bootstrap.py#L739 doesn't seem to reference the correct line of code anymore, as a rule of thumb it is better to copy the _permanent_ link (it creates a link to the current commit instead of to the current branch)
> 
> image

@Julien00859 PR description updated with permalinks

**Comment by rousseldenis** [comment]
> Shouldn't we remove the `find_module` bits too then?
> 
> Can you change the title to `[FIX] core: importlib find_module is deprecated` and elaborate on the commit message, explaining the rationals and giving relevant links to the documentation? I'm speaking about the commit message, not the PR message. Only the commit message is accessible within the git history. Also the link https://github.com/python/cpython/blob/main/Lib/importlib/_bootstrap.py#L739 doesn't seem to reference the correct line of code anymore, as a rule of thumb it is better to copy the _permanent_ link (it creates a link to the current commit instead of to the current branch)
> 
> image

@Julien00859 Done

**Comment by Julien00859** [comment]
looks like you need to rebase your branch too for the runbot
I'm reading the various things in the machinery hang thigh...

**Comment by rousseldenis** [comment]
@Julien00859 All :green_apple:

**Review by Julien00859 — APPROVED** [review]
Working as intended, the `load_module` method are no called anymore, the deprecation warnings work too and it is normal that the function returns nothing in case nothing was found, it is used as follow in the stdlib:

https://github.com/python/cpython/blob/aecf6aca515a203a823a87c711f15cbb82097c8b/Lib/importlib/_bootstrap.py#L1161-L1169
```py
    for finder in meta_path:
        with _ImportLockContext():
            try:
                find_spec = finder.find_spec
            except AttributeError:
                continue
            else:
                spec = find_spec(name, path, target)
        if spec is not None:
```

Is there any reason why to left the `load_module` methods? Couldn't we remove them?

**Comment by rousseldenis** [comment]
> Working as intended, the `load_module` method are no called anymore and the deprecation warnings work too. Is there any reason why to left the `load_module` methods? Couldn't we remove them?

@Julien00859 I admit I didn't explore more the subject and didn't removed them as I didn't know if some processes elsewhere would call them. This change works fine for our environment with python >= 3.10

**Comment by rousseldenis** [comment]
> I suggest we re-open this PR on master and that we remove the should-be-useless `find_module`, it'll give time to people to manifest that their odoo is broken and we'll have some time to restore the should-be-useless `find_module` in case it actually is not that useless

No problem. A backport, then ?

**Comment by Julien00859** [comment]
We can merge this one that just add find_spec and in the forward-port (in master) to remove find_module, yes let's do that insteat

@robodoo r+

**Pull request closed without merging**